Adding Differential Binarization model from PaddleOCR to Keras3

gowthamkpr opened the pull request:
This adds the Differentiable Binarization model for scene text detection. 

I implemented the architecture based on [ResNet50_vd](https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/algorithm_det_db_en.md) from PaddleOCR and ported the weights.

Demo colab: https://colab.research.google.com/gist/gowthamkpr/bd4a7f7742e92e66cfc57827052b8619/keras_paddleocr_v3.ipynb

mattdangerw commented:
Let's split this up. Start with ResNetVD backbone?

Some notes...

- Remove the aliases. One `ResNetVDBackbone` can handle all of these with different presets.
- Conversion scripts as scripts not colabs.
- Follow the local style for backbones as closely as possible. See some comments here https://github.com/keras-team/keras-nlp/pull/1737
- Keep models a flat directory. No `backbones/xx` etc.
- Add some tests.

divyashreepathihalli commented:
@gowthamkpr  is the PR ready for review?

divyashreepathihalli commented on keras_nlp/src/models/diffbin/diffbin.py:
```diff
@@ -0,0 +1,243 @@
+# Copyright 2024 The KerasNLP Authors
```
rename folder to `differential_binarization` and  file to `differential_binarization.py`

divyashreepathihalli commented on keras_nlp/src/models/diffbin/diffbin.py:
```diff
@@ -0,0 +1,243 @@
+        backbone = backbone
+
+        inputs = backbone.input
+        x = backbone.pyramid_outputs
```
please create a file `differential_binarization_backbone.py` and move the `diffbin_fpn_model` and backbone code into that. You can rename the `backbone` you are using in this file to `image_encoder` in the `differential_binarization_backbone` file. The task model should contain the preprocessor, backbone and the task `head`.

divyashreepathihalli commented on keras_nlp/src/models/diffbin/losses.py:
```diff
@@ -0,0 +1,139 @@
+from keras import ops
+
+
+class DiceLoss:
```
add test coverage for the losses here

divyashreepathihalli reviewed (commented):
Thanks for the PR! I have left a reorganization comment.

example for structuring the code - https://github.com/keras-team/keras-hub/tree/master/keras_hub/src/models/sam

divyashreepathihalli commented:
Hi @gowthamkpr! can you please refactor the code to KerasHub style?

- [ ] Add a preprocessor flow
- [ ] subclass image segementer model for the task class
- [ ] add preset class
- [ ] add standard test routines

gowthamkpr commented:
> Hi @gowthamkpr! can you please refactor the code to KerasHub style?

I've refactored using SAM as example.

> 
>     * [ ]  Add a preprocessor flow

I've added `DifferentialBinarizationPreprocessor` and `DifferentialBinarizationImageConverter`.

> 
>     * [ ]  subclass image segementer model for the task class

I've subclassed `ImageSegmenter`, but I left the custom `compile()` method, since we need a different loss than the one used in `ImageSegmenter`'s `compile()`.

> 
>     * [ ]  add preset class

Done. The model is not yet in Kaggle, so I've disabled the presets test for now.

> 
>     * [ ]  add standard test routines

Done. Not sure if there are additional standard test routines other than the ones used in SAM that should be run.

divyashreepathihalli reviewed (commented):
Thanks Gowtham! left a few comments!

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization_backbone.py:
```diff
@@ -0,0 +1,103 @@
+        fpn_channels=256,
+        **kwargs,
+    ):
+        inputs = image_encoder.input
```
Done. The code mixes the functional model with layer creation, similar to e.g. ResNet. Therefore we don't have a "=== Layers ===" section here.

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization_test.py:
```diff
@@ -0,0 +1,70 @@
+
+class DifferentialBinarizationTest(TestCase):
+    def setUp(self):
+        self.images = ops.ones((2, 224, 224, 3))
```
Done.

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization_test.py:
```diff
@@ -0,0 +1,70 @@
+        image_encoder = ResNetBackbone(
+            input_conv_filters=[64],
+            input_conv_kernel_sizes=[7],
+            stackwise_num_filters=[64, 128, 256, 512],
```
Done

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization.py:
```diff
@@ -0,0 +1,189 @@
+            instance.
+        head_kernel_list: list of ints. The number of filters for probability
+            and threshold maps. Defaults to [3, 2, 2].
+        step_function_k: float. `k` parameter used within the differential
```
Removed

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization.py:
```diff
@@ -0,0 +1,189 @@
+    Args:
+        backbone: A `keras_hub.models.DifferentialBinarizationBackbone`
+            instance.
+        head_kernel_list: list of ints. The number of filters for probability
```
Done.

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization_backbone_test.py:
```diff
@@ -0,0 +1,45 @@
+                56,
+                256,
+            ),
+            run_mixed_precision_check=False,
```
No. I tried adding an explicit `dtype` argument, but the problem remains that the mixed precision check checks against each sublayer of the model. The ResNet backbone, which is instantiated separately, therefore has the wrong `dtype`.

divyashreepathihalli commented on keras_hub/src/models/differential_binarization/differential_binarization_backbone.py:
```diff
@@ -0,0 +1,197 @@
+    pyramid network.
+
+    Args:
+        image_encoder: A `keras_hub.models.ResNetBackbone` instance.
```
add all args in docstring

gowthamkpr replied:
Great catch. Done.

divyashreepathihalli commented on keras_hub/src/models/differential_binarization/differential_binarization_backbone.py:
```diff
@@ -0,0 +1,197 @@
+
+
+def diffbin_fpn_model(inputs, out_channels, dtype=None):
+    in2 = layers.Conv2D(
```
what is in2 - can we rename this to be more readable?

gowthamkpr replied:
Yes, good point. I renamed the variables and added comments to make it better understandable.

divyashreepathihalli commented on keras_hub/src/models/differential_binarization/differential_binarization_image_converter.py:
```diff
@@ -0,0 +1,10 @@
+
+@keras_hub_export("keras_hub.layers.DifferentialBinarizationImageConverter")
+class DifferentialBinarizationImageConverter(ImageConverter):
+    backbone_cls = DifferentialBinarizationBackbone
```
there should be some resizing/rescaling ops here right?

gowthamkpr replied:
Depends. Basically these image operations are implemented in the super class, `ImageConverter`, and can be used as depicted in the demo colab I've added in the PR description. Dedicated code in this class might make sense to resize to resolutions of multiples of 32, which the model requires. On the other hand, it might be confusing for the user if the masks that are predicted have different resolutions than the input.

divyashreepathihalli replied:
you might want to look into Segformer for this. The output masks will need to be resized as well

divyashreepathihalli commented on keras_hub/src/models/differential_binarization/differential_binarization_ocr.py:
```diff
@@ -0,0 +1,113 @@
+
+
+@keras_hub_export("keras_hub.models.DifferentialBinarizationOCR")
+class DifferentialBinarizationOCR(ImageSegmenter):
```
we need to add a new base class for ocr, I don't think ImageSegmenter is a good. one. Do you have a specific reason you chose to subclass ImageSegmenter?

gowthamkpr replied:
Actually you suggested to subclass `ImageSegmenter` ([here](https://github.com/keras-team/keras-hub/pull/1739#issuecomment-2436214631)) if I understood correctly. Technically, the task is somewhat similar to segmentation tasks. We can of course add a separate base class for it to catch the semantic differences, but I would rather name it "scene text detection".

divyashreepathihalli replied:
Yeah I think it is better to add a new base class for OCR.

gowthamkpr replied:
Sure. I suggest to create an `ImageTextDetector` base class and include in this class the code (from the notebook) for translating the segmentation mask output into polygons, which are often needed in such applications. I'll try to get rid of the OpenCV and shapely dependencies in this code, I believe we don't have them in our requirements.
Does this work for you?

divyashreepathihalli replied:
sounds good!

mattdangerw replied:
What's the output of the task? Seems like this is not quite OCR right? Not text output?

Is this just a piece of what we would need for a full OCR setup?

gowthamkpr replied:
> What's the output of the task? Seems like this is not quite OCR right? Not text output?
> 
> Is this just a piece of what we would need for a full OCR setup?

It performs scene text detection, not OCR in the narrow sense. It belongs to OCR in a wider sense, though.

With text detection, we find where there is text in the image. The model typically outputs polygons or bounding boxes (after postproc), indicating the positions of text fragments. These portions of the image are then fed into an OCR model to get the text in ASCII.

gowthamkpr replied:
> Sure. I suggest to create an `ImageTextDetector` base class and include in this class the code (from the notebook) for translating the segmentation mask output into polygons, which are often needed in such applications. I'll try to get rid of the OpenCV and shapely dependencies in this code, I believe we don't have them in our requirements. Does this work for you?

I've added the `ImageTextDetector` task with the logic for transformation to polygons. With our dependencies, this requires quite a bit of code, but it works well.

mattdangerw replied:
@gowthamkpr what's the high level use journey we'd expect this to be used with? Is this eventually going to be part of an ocr pipeline? What would the full ocr pipeline code look like? E.g.

```python
ocr = keras_hub.OCR.from_preset("something_paddle_paddle")
ocr.predict(image) # returns text and polygons for text?
```

Or does this look like a two stage thing where we first go image -> polygons, and second go image and polygons -> text?

I think the thing we are missing is the high level look of the user journey we are trying to cover. Once we have that figured out, the rest of this will become a lot clearer.

gowthamkpr replied:
I would suggest exposing this two stage process to the user, i.e. image -> polygons and image+polygons -> text.

Differential Binarization is not able to perform the actual OCR, so we'll have to staple another model to it to perform the full pipeline. Additionally, it might be useful to have additional preprocessing steps in between, e.g., to rotate a region if a flat, but tilted bounding box has been detected. This is too much detail for hiding it from the user imho.

Maybe something like 
```python
# prediction
text_detector = keras_hub.OCR.from_preset("diffbin")
map_output = text_detector.predict(image)
# get polygons
polygons = text_detector.postprocess_to_polygons(map_output)

# OCR (not yet available)
ocr = keras_hub.OCR.from_preset("something")
for polygon in polygons:
  image_patch = keras_hub.utils.extract_patch(image, polygon)
  text = ocr.predict(image_patch)
```

mattdangerw replied:
I think the two stage process is fine if it's standard, but we might want to clean it up a bit.

gowthamkpr replied:
> Some rules for designing here.
> 
>     * We should aim for simple usage that stays close to regular Keras abstractions.
> 
>     * A task should definite a type of input and output data for any model that we build for that task. The goal is standardization, swap the model you are using without changing the high-level code using the model. So in this case, using OCR for two different types of inputs and outputs is not good.
> 
>     * Task class name is always `SomeModelSomeTask(SomeTask)`. So if the task is called `ImageTextExtractor` the subclass is `DiffBinImageTextExtractor`. If the task is `OCR` the subclass is `SomeModelOCR`. I don't think we need exceptions to this.
> 

Changed.


>     * What's the format of a polygon list. Do we keep it as a padded tensor?

At the moment I return a list(batch) of lists (image) of lists(polygon) of 2d tuples (points in a polygon). Considering this nested structure, and considering that the number vertices in a polygon might vary wildly, I think it makes sense to avoid forcing this structure into a tensor.

I'd have to check if it would even be possible to rewrite the contour finding algorithms to only use tensor operations. When using OpenCV as contour finder (which I currently optionally allow), it of course isn't.

>     * What's the format of a patch? It's not rectangular right? Is it a image plus a mask?

The text recognition model will accept a rectangular patch that only contains the text region. Within this patch, we could possibly mask out regions that are outside the polygon.

> 
>     * What will training or fine-tuning look like? What's the format of the input data?
>
Currently the segmentation masks need to be provided as input for fine-tuning, just as the model outputs segmentation masks before postprocessing. Changing this to accept a polygon representation for training should be possible, but then I'd prefer to override `train_step` for doing the conversion. Not sure how well this can be performed in the computation gradph, and I'm also not sure how sophisticated Keras' support for ragged tensors is to represent this nested structure of points.
 
> 
> Seems like our main options are, splitting detection and extraction as you are doing so far...
> 
> ```python
> # Inference.
> detector  = keras_hub.TextDetector.from_preset("diff_bin_small")
> # Some for of masking here so these can stay square tensors?
> polygons = detector.predict(images)
> extractor = keras_hub.TextExtractor.from_preset("other_thing")
> texts = extractor.predict({
>     "images": images,
>     "polygons": polygons,
> })
> 
> # Fine-tuning.
> # load some data. think this through! what is the format? do we need masking?
> images, polygons, texts = ...
> detector.fit(x=images, y=polygons)
> extractor.fit(x={"images": images, "polygons": polygons}, y=texts)
> ```
> 
> Or trying to bake this all into one.
> 
> ```python
> # Inference.
> ocr  = keras_hub.OCR.from_preset("diff_bin_and_other_thing")
> # Some for of masking here so these can stay square tensors?
> texts, polygons = ocr.predict(images)
> 
> # Fine-tuning. All at once? Split?
> ```

As I user, I'd have a strong preference for the split approach. Both model types are pretty complex, and I'd like to know what comes out of the first one, possibly identifying problems, before passing it on to the other model.

divyashreepathihalli reviewed (changes requested):
Thanks for the PR Gowtham! Left a few comments. Can you please also add a demo colab in the PR description to verify the model is working before merging?

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization_backbone.py:
```diff
@@ -0,0 +1,197 @@
+        )
+
+        outputs = {
+            "probability_maps": probability_maps,
```
They're different heads with the same architecture. `probability maps` predicts the mask where there is text in the image, `threshold_maps` predicts the border around these regions.

mattdangerw commented on keras_hub/src/models/differential_binarization/differential_binarization_ocr.py:
```diff
@@ -0,0 +1,113 @@
+@keras_hub_export("keras_hub.models.DifferentialBinarizationOCR")
+class DifferentialBinarizationOCR(ImageSegmenter):
+    """
+    A Keras model implementing the Differential Binarization
```
Same here, a one liner on the first line

mattdangerw commented on keras_hub/src/models/differential_binarization/differential_binarization_ocr.py:
```diff
@@ -0,0 +1,113 @@
+        backbone=backbone
+    )
+
+    detector(input_data)
```
What does the output of the task look like?

gowthamkpr replied:
We get the probability map, threshold map and binary map output in the last dimension from the model. I've added some documentation here.

mattdangerw commented on keras_hub/src/models/differential_binarization/differential_binarization_presets.py:
```diff
@@ -0,0 +1,17 @@
+"""Differential Binarization preset configurations."""
+
+backbone_presets = {
+    "diffbin_r50vd_icdar2015": {
```
I think we should use consistent abbreviations. If the class is `DifferentialBinarizationXX`, then the preset name would be `"differential_binarization"`. Or we could do `DiffBinBackbone` and `diff_bin_....` for the preset name.

Part of me is actually tempted by the latter, since `DifferentialBinarization` is such a mouthful. But I am not sure what is common in other libraries? Is there a common pattern in other projects?

gowthamkpr replied:
PaddleOCR mostly uses "DB", but I find this abbreviation somewhat indistinctive (in the sense you have a hard time googling it). I've also seen DBNet, but I don't know where this comes from, as the paper never names it DBNet, but just "Differential Binarization".

Agreed that `DifferentialBinarization` is quite long though and consistency is always good.

mattdangerw replied:
Let's go with DiffBin everywhere!

mattdangerw commented on keras_hub/src/models/differential_binarization/losses.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class DiceLoss:
+    def __init__(self, eps=1e-6, **kwargs):
```
A brief comment explaining these would be helpful.

gowthamkpr replied:
Done.

mattdangerw commented on keras_hub/src/models/differential_binarization/losses.py:
```diff
@@ -0,0 +1,126 @@
+
+
+class MaskL1Loss:
+    def __init__(self, **kwargs):
```
Same here, and so on.

mattdangerw commented on keras_hub/src/models/differential_binarization/losses.py:
```diff
@@ -0,0 +1,126 @@
+        return balance_loss
+
+
+class DBLoss(keras.losses.Loss):
```
Again lets be consistent in our abbreviations. We have DiffBin, DB and DifferentialBinarization. Maybe DiffBin is the right middle ground everywhere? But as I mentioned above, not sure what is common in other projects.

gowthamkpr replied:
(see above comment) I'll rename as soon as we have agreed on a good name.

mattdangerw reviewed (commented):
Mostly questions and some style stuff.

I am curious what to do with the task here. Does this output a segmentation mask? Maybe most importantly, does this fit into a bigger picture of an OCR system? If so, how do we expect the whole thing to work?

gowthamkpr commented on keras_hub/src/models/differential_binarization/differential_binarization_backbone.py:
```diff
@@ -0,0 +1,219 @@
+
+@keras_hub_export("keras_hub.models.DifferentialBinarizationBackbone")
+class DifferentialBinarizationBackbone(Backbone):
+    """
```
I've improved/added the docstrings here and in `losses.py`. ptal

mattdangerw commented on keras_hub/src/models/image_text_detector.py:
```diff
@@ -0,0 +1,277 @@
+
+    `ImageTextDetector` tasks wrap a `keras_hub.models.Task` and
+    a `keras_hub.models.Preprocessor` to create a model that can be used for
+    image segmentation.
```
this is false right? This task will detect polygons in an image?

alexhartl replied:
Sure. Thanks!

mattdangerw commented on keras_hub/src/models/image_text_detector.py:
```diff
@@ -0,0 +1,277 @@
+    can be used to load a pre-trained config and weights.
+
+    Args:
+        detection_thresh: The value for thresholding predicted mask outputs.
```
Are these general concepts that will apply to other models? We should only leave common stuff here. If this would not apply to a non-diff bin model, we should leave it out.

gowthamkpr replied:
It's pretty typical for such models to output a mask, which then needs to be translated into polygon form, but there are also models that follow different approaches. So as you prefer - we can move it into `diffbin_textdetector.py` or even create a separate file for the postproc.

mattdangerw commented on keras_hub/src/models/image_text_detector.py:
```diff
@@ -0,0 +1,277 @@
+        )
+        return config
+
+    def postprocess_to_polygons(self, masks, contour_finder="simple"):
```
We might want to consider rolling this into the preprocessing layer. That is what we do for generation.

gowthamkpr replied:
Not sure what you mean here. Could you please elaborate or provide pointers for how it is implemented for generation?

divyashreepathihalli commented on keras_hub/src/models/diffbin/diffbin_preprocessor.py:
```diff
@@ -0,0 +1,14 @@
+
+
+@keras_hub_export("keras_hub.models.DiffBinPreprocessor")
+class DiffBinPreprocessor(ImageSegmenterPreprocessor):
```
if this model is going to be returning polygons - we might need to change the parent class

gowthamkpr replied:
Yes. The functionality that we require here is pretty similar, though (resize the output mask in addition to the model's output). So basically, we'd have to copy  `ImageSegmenterPreprocessor`.

divyashreepathihalli reviewed (commented):
@gowthamkpr if you have addressed the comments can you please respond or resolve - so we can merge this PR

sachinprasadhs commented:
Closing this since the continuation PR is in progress here https://github.com/keras-team/keras-hub/pull/2095
